### PR TITLE
Use data sketches to build column stats in ANALYZE

### LINKED ISSUE #15152: Reduce amount of memory required for running `ANALYZE`
### Problem Statement

Currently `ANALYZE` gets `30_000` samples per table per node. See:

- https://github.com/crate/crate/blob/df549946d1774f2539e0ce1e6b7ef1bc60dbf298/server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java#L85
- https://github.com/crate/crate/blob/df549946d1774f2539e0ce1e6b7ef1bc60dbf298/server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java#L242-L249

That can use up a lot of memory and result in high GC activity.

### Possible Solutions

We create table stats from these samples:

https://github.com/crate/crate/blob/df549946d1774f2539e0ce1e6b7ef1bc60dbf298/server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java#L194-L223

Maybe we could already compress the samples in the node-action in some form, ideally incrementally, to never materialize `30_000` records, and then merge the stats together.
E.g.
- hyperloglog could be re-used to calculate the distinct values
- [frequency sketch](https://datasketches.apache.org/docs/Frequency/FrequencySketchesOverview.html) for the most common values could be an option?

### Considered Alternatives

- Configure or auto-tune sampling size based on available heap/record size, but that reduces the quality of the stats.
- Suffer the GC
- Throw hardware at the problem

### PR DESCRIPTION
**Pull request by romseygeek**
Rather than streaming selected rows between nodes to build column stats, we
instead build distinct, frequency and quantile sketches for individual columns
on each data nodes and stream these instead.  These sketches can be merged
efficiently to produce column stats without requiring large memory overheads
or creating many object allocations.

Fixes #15152

### CODE REVIEW

**Comment by romseygeek** [comment]
I haven't added a changes entry yet because we need to decide whether to target 5.7.0 or wait for the next release.

@matriv ran some benchmarks for the ANALYZE command against the latest version of this branch before I rebased, and got the following results:

```
Q: analyze
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      513.119 ±   13.300 |    498.485 |    512.163 |    514.420 |    684.115 |
|   V2    |      448.307 ±    7.683 |    445.518 |    447.230 |    447.948 |    551.800 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  13.48%                           -  13.54%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 13.48%
The test has statistical significance


System/JVM Metrics (durations in ms, byte-values in MB)
    |    YOUNG GC            |       OLD GC           |      HEAP         |     ALLOC     
    |  cnt      avg      max |  cnt      avg      max |  initial     used |     rate      total
 V1 |   70     2.87     2.75 |   19    14.69    13.73 |     4295       76 |  2053.72     209756
 V2 |   20     3.43     1.59 |    6    17.65    16.11 |     4295      129 |   940.77      84047
```

So analyze is slightly faster now (it was much faster before I added rate limiting back in!) and causes much less GC pressure.

**Comment by romseygeek** [comment]
There are still some follow-ups to do here:
* Merge `columnStatsSupport` into `storageSupport` - a slightly noisy refactoring, doesn't need to be done immediately
* Use lucene field metadata directly to build stats where possible in place of sampling - should be even faster and less memory intensive
* Improve reporting of array type columns - build sketches directly on individual array values rather than on the whole array
* For further in the future: store stats in a system index, schedule updates based on when tables are updated

File: server/src/main/java/io/crate/statistics/DistinctValuesSketch.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+     * Creates a new empty sketch
+     */
+    public static DistinctValuesSketch newSketch() {
+        return new DistinctValuesSketch() {
+
+            final UpdateSketch distinctSketch = UpdateSketch.builder().build();
+
+            @Override
+            public void update(String v) {
+                distinctSketch.update(v);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/statistics/DistinctValuesSketch.java:64]: Is this okay to go with String only here? The `UpdateSketch` class has overloads for the different types.

Or is this also related  to the TODO about having more type specific sketches?

File: server/src/main/java/io/crate/statistics/Samples.java [hunk 2]
```diff
@@ -26,129 +26,87 @@
...
         this.numTotalDocs = numTotalDocs;
         this.numTotalSizeInBytes = numTotalSizeInBytes;
     }
 
-    @SuppressWarnings("rawtypes")
-    public Samples(List<Streamer> recordStreamer, StreamInput in) throws IOException {
-        this.recordStreamer = recordStreamer;
+    public Samples(List<Reference> references, StreamInput in) throws IOException {
+        if (in.getVersion().before(Version.V_5_7_0)) {
+            throw new UnsupportedOperationException("Cannot run ANALYZE in a mixed version cluster");
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/statistics/Samples.java:54]: Does this need some special handling in `TableStatsService` ?

File: server/src/test/java/io/crate/statistics/StatsUtils.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    public static <T> ColumnStats<T> statsFromValues(DataType<T> dataType, Collection<T> values, int nullCount) {
+        ColumnSketchBuilder<T> builder = dataType.columnStatsSupport().sketchBuilder();
+        builder.addAll(values);
+        for (int i = 0; i < nullCount; i++) {
+            builder.add(null);
+        }
+        return builder.toStats();
+    }
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/statistics/StatsUtils.java:43]: How about a `toStats(values)` on `ColumnStatsSupport` ? It would avoid another indirection and I think help a bit with the discoverability.

**Review by mfussenegger — APPROVED** [review]
Looks pretty good to me:


File reads before:

image

File reads after:

image


memory allocations before:

image


memory allocations after:

image

File: server/src/main/java/io/crate/statistics/DistinctValuesSketch.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/DistinctValuesSketch.java:64]: We can specialise this when we get numeric sketches, yes.  But at the moment the input could be any object, so we need to convert it to something that works for any type.

File: server/src/main/java/io/crate/statistics/Samples.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/Samples.java:54]: Oh that's a good idea, we can add a line to `update` that returns early if the minimum node version in the cluster is before 5.7.  I'll push an update.

File: server/src/test/java/io/crate/statistics/StatsUtils.java (hunk 3 above)
Comment (mfussenegger) [reply server/src/test/java/io/crate/statistics/StatsUtils.java:43]: Oh I missed that this is in the test sources. Agree that in that case it's better to keep as is.

File: server/src/main/java/io/crate/statistics/ColumnSketchBuilder.java [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import io.crate.Streamer;
+import io.crate.types.DataType;
+
+/**
+ * Constructs a {@link ColumnStats} from a set of samples
+ */
+public abstract class ColumnSketchBuilder<T> {
```
Comment (matriv) [inline server/src/main/java/io/crate/statistics/ColumnSketchBuilder.java:41]: same?

File: server/src/test/java/io/crate/integrationtests/AnalyzeITest.java [hunk 5]
```diff
@@ -56,9 +56,10 @@ public void test_analyze_statement_refreshes_table_stats_and_stats_are_visible_i
         assertThat(((Float) row[0]).doubleValue()).isCloseTo(0.166, withPercentage(1));
         assertThat(row[1]).isEqualTo(DataTypes.INTEGER.fixedSize());
         assertThat(row[2]).isEqualTo(3.0f);
-        assertThat(((List<String>) row[3])).isEmpty();
-        assertThat(((List<Double>) row[4])).isEmpty();
-        assertThat(((List<String>) row[5])).containsExactly("1", "2", "3");
+        assertThat(((List<String>) row[3])).containsExactly("3", "2", "1");
```
Comment (matriv) [inline server/src/test/java/io/crate/integrationtests/AnalyzeITest.java:59]: Doesn't matter I guess that the order is inverted?

File: server/src/test/java/io/crate/integrationtests/AnalyzeITest.java [hunk 6]
```diff
@@ -56,9 +56,10 @@ public void test_analyze_statement_refreshes_table_stats_and_stats_are_visible_i
...
+        assertThat(((List<Float>) row[4])).hasSize(3);
+        assertThat(((List<Float>) row[4]).get(0)).isEqualTo(0.5f);
+        assertThat(((List<String>) row[5])).isEmpty();
```
Comment (matriv) [inline server/src/test/java/io/crate/integrationtests/AnalyzeITest.java:62]: maybe I'm missing something but why `histogra_bounds` is empty now?

File: server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+import org.apache.datasketches.frequencies.ErrorType;
+import org.apache.datasketches.frequencies.ItemsSketch;
+import org.apache.datasketches.memory.Memory;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import io.crate.Streamer;
+
+public class MostCommonValuesSketch<T> {
```
Comment (matriv) [inline server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java:38]: same

File: server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            freqs[i] = (double) counts[i] / valueCount;
+        }
+        return freqs;
+    }
+
+    private static int cutoff(long valueCount, double approxDistinct, long[] counts) {
+        int cutoff = counts.length;
+
+        long sumFreq = 0;
+        for (int i = 0; i < counts.length - 1; i++) {
```
Comment (matriv) [inline server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java:105]: is `- 1` correct here?

File: server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        var otherSketch = (ItemsSketch<T>) other.sketch;
+        this.sketch.merge(otherSketch);
+        return this;
+    }
+
+    public void update(T value) {
+        this.sketch.update(value);
+    }
+
+    public MostCommonValues<T> toMostCommonValues(long valueCount, double approxDistinct) {
```
Comment (matriv) [inline server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java:71]: is it safe that `valueCount` is > 0 as is used in division?
(also for the methods below).

File: server/src/main/java/io/crate/statistics/SketchRamAccounting.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        this.rateLimiter = rateLimiter;
+    }
+
+    public void addBytes(long bytes) {
+        this.blockCache += bytes;
+        while (this.blockCache > BLOCK_SIZE) {
+            this.blockCache -= BLOCK_SIZE;
+            ramAccounting.addBytes(SHIFTED_BLOCK_SIZE);
+        }
+        checkRateLimit(bytes);
```
Comment (matriv) [inline server/src/main/java/io/crate/statistics/SketchRamAccounting.java:60]: As followup, does it make sense to not call `checkRateLimit` everytime, but rather have a counter to call this every `X` calls of `addBytes`?

**Review by matriv — COMMENTED** [review]
Thank you @romseygeek!
Left some comments and questions, apologies if for some I'm missing some context.

File: server/src/main/java/io/crate/statistics/DistinctValuesSketch.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+import org.apache.datasketches.theta.Sketch;
+import org.apache.datasketches.theta.Sketches;
+import org.apache.datasketches.theta.Union;
+import org.apache.datasketches.theta.UpdateSketch;
+import org.elasticsearch.common.io.stream.StreamInput;
+
+/**
+ * A streamable representation of a distinct values sketch
+ */
+public abstract class DistinctValuesSketch {
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/DistinctValuesSketch.java:37]: I did this initially, but ended up reverting it; these classes are only ever seen by their containing ColumnSketchBuilder, so having the extra interface just adds clutter.

File: server/src/test/java/io/crate/integrationtests/AnalyzeITest.java (hunk 5 above)
Comment (romseygeek) [reply server/src/test/java/io/crate/integrationtests/AnalyzeITest.java:59]: This was previously the histogram bounds, but we're getting the most common values back instead (which are always in frequency order).  I'm not sure why it wasn't returning the most common values before, to be honest - it's a small enough set that it should have always been returning them?  And histogram bounds don't contain any value that appears in MCV, so if you return all values as part of MCV then histogram bounds will be empty.

File: server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java (hunk 8 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java:105]: It is not!  Good catch.

Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java:105]: Oh no, hang on, it is correct.  We are calculating the sum of frequencies *excluding* the last value, to see if the frequency of the last value is low enough in comparison that it can be excluded.  I'll add a comment.

Comment (matriv) [reply server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java:105]: :+1 thx!

File: server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java (hunk 9 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/MostCommonValuesSketch.java:71]: It is, because if `valueCount` is zero then `approxDistinct` will also be zero and so we exit early.  I'll add some tests that explicitly check empty sketch builders.

File: server/src/main/java/io/crate/statistics/SketchRamAccounting.java (hunk 10 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/SketchRamAccounting.java:60]: We could move the rate limit check inside the if statement above to make it check every 1024 bytes instead?

Comment (matriv) [reply server/src/main/java/io/crate/statistics/SketchRamAccounting.java:60]: Also, yes, just avoid some if branches and the += bytes on every value processed.

File: server/src/main/java/io/crate/statistics/SketchStreamer.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+import org.apache.datasketches.common.ArrayOfItemsSerDe;
+import org.apache.datasketches.memory.Buffer;
+import org.apache.datasketches.memory.Memory;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.StreamInput;
+
+import io.crate.Streamer;
+
+class SketchStreamer<T> extends ArrayOfItemsSerDe<T> {
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/SketchStreamer.java:38]: It's indirectly tested in the integration tests, but I can add some tests that explicitly check that the sketches can be streamed and reconstituted.

[the remark this answers is not in the document]
Comment (matriv) [reply server/src/main/java/io/crate/statistics/SketchStreamer.java:38]: Yes please, If something changes in the future, streaming tests are valuable, otherwise you can bang your head for long hours to figure out what's breaking.

**Comment by romseygeek** [comment]
@matriv have pushed changes addressing your feedback

**Comment by romseygeek** [comment]
@Mergifyio squash title+body

**Review by matriv — APPROVED** [review]
Nice work, LGTM!
